lib/inv_cg_quda.cpp

```diff
@@ -688,20 +586,22 @@ namespace quda {
 
       pAp = blas::reDotProduct(p, Ap);
 
-      alpha = r2 / pAp;
+      for (auto i = 0u; i < alpha.size(); i++) alpha[i] = r2[i] / pAp[i];
 
       // here we are deploying the alternative beta computation
-      double2 cg_norm = blas::axpyCGNorm(-alpha, Ap, rSloppy);
-      r2 = cg_norm.x;                            // (r_new, r_new)
-      sigma = cg_norm.y >= 0.0 ? cg_norm.y : r2; // use r2 if (r_k+1, r_k+1-r_k) breaks
-      rNorm = sqrt(r2);
+      auto cg_norm = blas::axpyCGNorm(-alpha, Ap, r_sloppy);
+      for (auto i = 0u; i < cg_norm.size(); i++) {
+        r2[i] = cg_norm[i].x;                                  // (r_new, r_new)
+        sigma[i] = cg_norm[i].y >= 0.0 ? cg_norm[i].y : r2[i]; // use r2 if (r_k+1, r_k+1-r_k) breaks
+      }
+      rNorm = sqrt(r2[0]);
 
       // If the iterated norm has dropped by more than a factor of delta, trigger
       // an update. The baseline we check against differs depending on if
       // we're still checking the L2 norm, or if that has converged/broken down and we're
       // now looking at the HQ residual.
 
-      if (!L2breakdown && (L2_required || convergenceL2(r2, hq_res, stop, param.tol_hq))) {
+      if (!L2breakdown && (L2_required || convergenceL2(hq_res, param.tol_hq))) {
```

Comment:
This should be `convergenceL2(r2, stop)` not `convergenceL2(hq_res, param.tol_hq)`

Reply:
50e78799c7b9e5babb2ced06bcf2f27e35f036e3